src/joystick/gdk/SDL_gameinputjoystick.cpp

```diff
@@ -0,0 +1,592 @@
+    /* SDL expects a nanosecond timestamp, so I guess US_TO_NS should be used here? */
+    ts = SDL_US_TO_NS(reading->GetTimestamp());
+
+    if (((!hwdata->lastTimestamp) || (ts != hwdata->lastTimestamp)) && reading->GetGamepadState(&state)) {
+        /* `state` is now valid */
+
+#define tosint16(_TheValue) ((Sint16)(((_TheValue) < 0.0f) ? ((_TheValue) * 32768.0f) : ((_TheValue) * 32767.0f)))
+        SDL_SendJoystickAxis(ts, joystick, 0, tosint16(state.leftThumbstickX));
+        SDL_SendJoystickAxis(ts, joystick, 1, tosint16(state.leftThumbstickY));
+        SDL_SendJoystickAxis(ts, joystick, 2, tosint16(state.leftTrigger));
+        SDL_SendJoystickAxis(ts, joystick, 3, tosint16(state.rightThumbstickX));
+        SDL_SendJoystickAxis(ts, joystick, 4, tosint16(state.rightThumbstickY));
+        SDL_SendJoystickAxis(ts, joystick, 5, tosint16(state.rightTrigger));
+#undef tosint16
+
+        for (btnidx = 0; btnidx < (Uint8)SDL_arraysize(s_XInputButtons); ++btnidx) {
+            if (s_XInputButtons[btnidx] == 0) {
+                btnstate = SDL_RELEASED;
+            } else {
+                btnstate = (state.buttons & s_XInputButtons[btnidx]) ? SDL_PRESSED : SDL_RELEASED;
+            }
+
+            SDL_SendJoystickButton(ts, joystick, btnidx, btnstate);
+        }
+
+        if (state.buttons & GameInputGamepadDPadUp) {
+            hat |= SDL_HAT_UP;
+        }
+        if (state.buttons & GameInputGamepadDPadDown) {
+            hat |= SDL_HAT_DOWN;
+        }
+        if (state.buttons & GameInputGamepadDPadLeft) {
+            hat |= SDL_HAT_LEFT;
+        }
+        if (state.buttons & GameInputGamepadDPadRight) {
+            hat |= SDL_HAT_RIGHT;
+        }
+        SDL_SendJoystickHat(ts, joystick, 0, hat);
+
+        /* Xbox doesn't let you obtain the power level, pretend we're always full */
+        SDL_SendJoystickBatteryLevel(joystick, SDL_JOYSTICK_POWER_FULL);
+
+        hwdata->lastTimestamp = ts;
+    }
+
+    reading->Release();
+}
+
+static void GAMEINPUT_JoystickClose(SDL_Joystick* joystick)
+{
+    SDL_free(joystick->hwdata);
+    joystick->hwdata = NULL;
+}
+
+static void GAMEINPUT_JoystickQuit(void)
+{
+    int idx;
+
+    if (!g_pGameInput) {
+        return;
+    }
+
+    /* free the callback */
+    g_pGameInput->UnregisterCallback(g_GameInputCallbackToken, /*timeoutInUs:*/ 10000);
+    g_GameInputCallbackToken = GAMEINPUT_INVALID_CALLBACK_TOKEN_VALUE;
+
+    /* free the list */
+    for (idx = 0; idx < g_GameInputList.count; ++idx) {
+        g_GameInputList.devices[idx]->device->Release();
+        SDL_free(g_GameInputList.devices[idx]);
+        g_GameInputList.devices[idx] = NULL;
+    }
+    SDL_free(g_GameInputList.devices);
+    g_GameInputList.count = 0;
+
+    g_pGameInput->Release();
+    g_pGameInput = NULL;
+}
+
+static SDL_bool GAMEINPUT_JoystickGetGamepadMapping(int device_index, SDL_GamepadMapping *out)
+{
+    return SDL_FALSE;
```

Comment:
You can create the automatic mapping here. This can be added later after your code is merged though. If I can build this using public tools then I can implement this pretty quickly.

Reply:
I used it in my previous attempt, but it's lacking some stuff the strings have, like axis inversion (though I can probably just invert the axis in the backend itself, we're simulating anyway)